k8srm-prototype/pkg/api/claim_types.go

```diff
@@ -126,42 +74,95 @@ type DeviceClaim struct {
+// or a group for which one must be sastisfied.
+type DeviceClaimInstance struct {
+	// At least one of AllOf and OneOf must be populated.
+
+	// If fields of DeviceClaimDetail are populated, OneOf should
+	// be empty.
+	DeviceClaimDetail `json:",inline"`
+
+	// OneOf contains a list of claims, only one of which must be satisfied.
+	// Claims are listed in order of priority.
+	//
+	// +optional
+	OneOf []DeviceClaimDetail `json:"oneOf,omitempty"`
+}
+
+// DeviceClaimDetail contains the details of how to fulfill a specific
+// request for devices.
+type DeviceClaimDetail struct {
+	// DeviceType may be specified to get a device from any class that
+	// supports this type of device. For example, the user can request
+	// an 'sriov-nic', and any class that can provide that type will be
+	// considered for fulfillment of the claim.
+	//
 	// +optional
-	Driver *string `json:"driver,omitempty"`
+	DeviceType *string `json:"deviceType"`
+
+	// DeviceClass is the name of the DeviceClass containing the basic information
+	// about the device being requested.
+	//
+	// +optional
+	DeviceClass *string `json:"deviceClass"`
 
 	// Constraints is a CEL expression that operates on device attributes.
 	// In order for a device to be considered, this CEL expression and the
 	// Constraints expression from the DeviceClass must both be true.
+	//
 	// +optional
 	Constraints *string `json:"constraints,omitempty"`
 
-	// Device classes and claims may represent or be satisfied by choosing
-	// multiple devices instead of just a single device.
+	// Device claims may be satisfied by choosing multiple devices instead
+	// of just a single device. How and when that happens depends on the
+	// requests and limits specified.
 
-	// MinDeviceCount is the minimum number of devices that should be selected
-	// for this claim. It must be greater than or equal to the class MinDeviceCount,
-	// and less than or equal to the class MaxDeviceCount. Default is 1.
-	// +optional
-	MinDeviceCount *int `json:"minDeviceCount,omitempty"`
+	// Requests allows the user to specify the minimum requirements that
+	// must be satisfied across all devices allocated for this claim.  All
+	// drivers can support "count" for requests, but other resource types
+	// are driver-specific. The default value for requests is a single
+	// entry for "count" with value of 1.
+	Requests map[string]resource.Quantity `json:"requests,omitempty"`
 
-	// MaxDeviceCount is the maximum number of devices that should be selected
-	// for this claim. It must be less than or equal to the class MaxDeviceCount.
-	// Default is no maximum.
-	// +optional
-	MaxDeviceCount *int `json:"maxDeviceCount,omitempty"`
+	// Limits allows the user to control the maximum count of devices
+	// that is allocated to satisfy the claim. Depending on the driver
+	// and device other resource limits may or may not be enforceable.
+	Limits map[string]resource.Quantity `json:"limits,omitempty"`
```

Comment:
If I could go back in time I might argue that pod resources "limits" was a mistake.  It seems like the exception, rather than the norm, for workloads to be allowed to access more than their request.

I think we can do better, will think.

Reply:
After chewing a bit, I don't see how `limits` applies to anything but count, so I think I prefer `MaxDeviceCount` (or something like that), but I think I would make it default to 1.  The normal case is that I expect a claim to be satisfied by one device and I need to explicitly allow the system to smear a request across 2 or more devices.